### ISSUE TITLE
What is the point of "nonconformity_quantiles" when "quantiles" can be specified independently at inference?

### ISSUE THREAD

**Issue by muhlbach** [radix-ai/conformal-tights#45]
Hi maintainers! 

Love your work! I'm curious if you could explain what the argument `nonconformity_quantiles` does and if it's important to get could predictions of the quantiles? How does it relate to the predict argument `quantiles`?

```
from conformal_tights import ConformalCoherentQuantileRegressor
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
from xgboost import XGBRegressor

X, y = make_regression(n_samples=1000, n_features=10, noise=0.1)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15, random_state=42)

my_regressor = XGBRegressor(objective="reg:absoluteerror")
conformal_predictor = ConformalCoherentQuantileRegressor(estimator=my_regressor, nonconformity_quantiles=[0.1, 0.9]) # <-- What does this do?
conformal_predictor.fit(X_train, y_train)

y_pred = conformal_predictor.predict(X_test)
y_pred_quantiles = conformal_predictor.predict_quantiles(
    X_test, quantiles=[0.3, 0.4, 0.5, 0.6, 0.7] # <-- When this can be specified
)
```

**Comment by lsorber** [comment]
Hi @muhlbach, thank you for the kind words!

> I'm curious if you could explain what the argument `nonconformity_quantiles` does and if it's important to get could predictions of the quantiles?

The `nonconformity_quantiles` are used to train a quantile regression model that predicts certain quantiles called _nonconformity quantiles_. You can think of these nonconformity quantiles as a rough measure of the uncertainty on a given prediction. However, they are troublesome in several ways:
1. They are not the quantiles you actually want to predict
2. They are not coherent: for example the 25th percentile prediction may be larger than the 75th percentile prediction (because quantile regression models for different quantiles are trained independently of each other)
3. They are not (conformally) calibrated: on test data the quantiles are unlikely to be accurate

Conformal Tights takes these rough measures of uncertainty as input, and turns them into output quantile predictions that do satisfy the properties you want:
1. The output quantiles you want
2. The output quantiles are coherent
3. The output quantiles are conformally calibrated

Here, we first achieve (1) and (2) through [Coherent Linear Quantile Regression](https://github.com/superlinear-ai/conformal-tights/blob/main/src/conformal_tights/_coherent_linear_quantile_regressor.py). Then, we achieve (3) through [Conformal Calibration](https://github.com/superlinear-ai/conformal-tights/blob/main/src/conformal_tights/_conformal_coherent_quantile_regressor.py).

> How does it relate to the predict argument `quantiles`?

The `quantiles` argument represents the quantiles you want to predict, while the `nonconformity_quantiles` are the 'raw materials' used to produce the requested quantiles.

Hope that helps! We'll make a note of improving the docstrings to clarify the above for other users too.

**Comment by muhlbach** [comment]
This is an excellent explanation! 
As a follow-up question, does this mean that in order to use conformal prediction to predict distribution I do need a quantile predictor and cannot use a general estimator?

**Comment by lsorber** [comment]
> This is an excellent explanation!

Happy to hear it was helpful!

> As a follow-up question, does this mean that in order to use conformal prediction to predict distribution I do need a quantile predictor and cannot use a general estimator?

No, it doesn't necessarily have to be a set of quantile predictors. The only requirement on the `nonconformity_estimator` is that it produces one or more 'nonconformity indicators' that together represent some indication of the variance (or more generally, of the distribution) of the point prediction.

A good candidate for such nonconformity indicators is indeed a set of quantile predictions corresponding to a set of nonconformity quantiles. But if you have another source of nonconformity indicators, such as a Bayesian estimate of the variance in a SVM model for example, it's absolutely fine to use that instead.

**Comment by muhlbach** [comment]
Thanks for your excellent guidance!
Last, other than the default settings, do you have some general rule-of-thumbs parameters that you use, e.g., what are good choices of `nonconformity_quantiles` and `quantiles`?

**Comment by lsorber** [comment]
> Thanks for your excellent guidance! Last, other than the default settings, do you have some general rule-of-thumbs parameters that you use, e.g., what are good choices of `nonconformity_quantiles` and `quantiles`?

- `nonconformity_quantiles`: I recommend sticking with the default value of `(1/8, 1/4, 1/2, 3/4, 7/8)`. These are the 'raw materials' that are calibrated and made coherent later on. You want enough of these to be able to predict accurate quantiles, but you don't want too many either because that increases the number of parameters and therefore the risk of overfitting. The 5 default quantiles strike a pretty good balance between these two objectives.
- `quantiles`: These are the quantiles you want to predict, and is completely up to you. That said, the computational cost tends to increase quite quickly as you add more quantiles. If you only need two quantiles and are more interested in attaining a minimum coverage, then I recommend setting `priority="coverage"` as well (the default is `priority="accuracy"`, which prioritises quantile accuracy over prediction interval coverage).

**Comment by muhlbach** [comment]
Thanks, this was very helpful!
